server/server/src/static_files.rs

```diff
@@ -70,3 +89,127 @@ async fn plugins(
+
+    while let Some(mut field) = payload.try_next().await? {
+        // A multipart/form-data stream has to contain `content_disposition`
+        let content_disposition = field.content_disposition();
+        log::info!(
+            "Uploading File: {}",
+            content_disposition.get_filename().unwrap_or_default()
+        );
+        log::debug!("Content Disposition: {:?}", content_disposition);
+        log::debug!("Content Type: {:?}", field.content_type());
+
+        let sanitized_filename =
+            sanitize_filename::sanitize(content_disposition.get_filename().unwrap_or_default());
+        let static_file = service
+            .reserve_file(&sanitized_filename, &file_category)
+            .map_err(|err| InternalError::new(err, StatusCode::INTERNAL_SERVER_ERROR))?;
+
+        files.push(UploadedFile {
+            id: static_file.id.clone(),
+            name: content_disposition
+                .get_name()
+                .unwrap_or_default()
+                .to_string(),
+            filename: content_disposition
+                .get_filename()
+                .unwrap_or_default()
+                .to_string(),
+            mime_type: field
+                .content_type()
+                .map(|mime| mime.to_string())
+                .unwrap_or_default(),
+        });
+
+        // File::create is blocking operation, use threadpool
+        let mut f = web::block(|| std::fs::File::create(static_file.path)).await??;
+
+        // Field in turn is stream of *Bytes* object
+        while let Some(chunk) = field.try_next().await? {
+            // filesystem operations are blocking, we have to use threadpool
+            f = web::block(move || f.write_all(&chunk).map(|_| f)).await??;
+        }
+    }
+    Ok(files)
+}
+
+async fn upload_sync_file(
+    payload: Multipart,
+    settings: Data<Settings>,
+    service_provider: Data<ServiceProvider>,
+    path: web::Path<(String, String)>,
+) -> Result<HttpResponse, Error> {
+    let db_connection = service_provider
+        .connection()
+        .map_err(|err| InternalError::new(err, StatusCode::INTERNAL_SERVER_ERROR))?;
+
+    let (table_name, record_id) = path.into_inner();
+
+    let files = handle_file_upload(
+        payload,
+        settings,
+        StaticFileCategory::SyncFile(table_name.clone(), record_id.clone()),
+    )
+    .await?;
+
+    let repo = SyncFileReferenceRowRepository::new(&db_connection);
+    for file in files.clone() {
+        let result = repo.upsert_one(&SyncFileReferenceRow {
+            id: file.id,
+            file_name: file.filename,
+            table_name: table_name.clone(),
+            // content_type: file.mime_type,
+            created_datetime: chrono::Utc::now().naive_utc(),
+            deleted_datetime: None,
+            record_id: record_id.clone(),
+        });
+        match result {
+            Ok(_) => {}
+            Err(err) => {
+                // TODO: delete the files that were uploaded?
```

Comment:
What should we do here?

Reply:
I think (hard) delete the uploaded - otherwise the user is encouraged to upload again, and then what happens? 
Ideally, the response is a list of successes and failures so the client can determine what to do.. but if we throw a server error and it's all or nothing, I think just clear the decks and reset the state to the original. ( I think the error is pretty unlikely so it's fine to delete )